### ISSUE TITLE
Overwriting a package instead of downloading it.

### ISSUE THREAD

**Issue by rbozan** [echogarden-project/echogarden#43]
Thanks for the package.

I'm wondering if there's a possibility to overwrite the supplied packages by something like a configuration file or environment variables. This is because something like sox would need patching if you are using NixOS. Eg using something like:

```sh
 patchelf \
  --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
  ~/.local/share/echogarden/packages/sox-14.4.2-linux-minimal-20230802/sox
```

However it'd be nice if we could just use the `sox` binary provided by the package manager, like how it's done in MacOS:

https://github.com/echogarden-project/echogarden/blob/c2f6bb4573b9bf375989550645841eb6c62a913e/src/audio/SoxPath.ts#L16-L21

This way this binary also would not need to download sox itself, as that can be done beforehand.

**Comment by rotemdan** [comment]
The thing is that I had so many problems with bugs with SoX, especially on Windows and macOS, that it was comparably so much better to just go and download a particular version I know should work, than dealing with whatever the user has installed. In that sense, it makes it more reproducible.

The reason I disabled the downloaded SoX package on macOS is because of security and signature issues. Since I don't have access to a macOS computer, I couldn't find a precompiled signed version of SoX (the one I had initially produced security errors). I tried really really hard (looking into things like Homebrew), but eventually gave up.

You are suggesting to provide an option to force the usage of whatever version of SoX the user has in their path. That's okay. Need to see how to do that since it's a global option not specific to a particular engine. Probably not via an environment variable though! That's something I'm trying to avoid.

Anyway, the only thing that SoX is used for is for audio playback. If you import Echogarden as a library, SoX is never actually used at all (though FFMpeg is still used for codec conversions).

**Comment by rbozan** [comment]
Ah I see. I'm just playing with the library at the moment, so I played around with the CLI. I indeed noticed I did not need the sox library when I supply a --language like so:

```
npx echogarden align ./visibilizar.mp3 ./transcript.txt ./result.json  --language=ES
```

-----------------

Another annoying point of the library at the moment, and looking at the commit history it seems intentional,  is that the library does not have a `package-lock.json` or shrinkwrap file. It makes it annoying to use because using Nix you'd have something like the following to build this package:

```nix
let echogardenPackage = pkgs.buildNpmPackage
  rec {
    pname = "echogarden";
    version = "0.11.22";

    src = pkgs.fetchFromGitHub {
      owner = "echogarden-project";
      repo = pname;
      rev = "main";
      hash = "sha256-Zkzhlu+qE6N+73D7XzARmGq9o3iBk4ZWvIJx1Lw2FkM=";
    };

    npmDepsHash = "sha256-dpegS7YboLjxDJ0uOfKAFkfQkI7cWp1PjpoCBmEKvBs=";

    # dontNpmBuild = true;
    npmFlags = [ "--legacy-peer-deps" ];

 };
```

Afterwards you can use it in your project or publish it as a [Nix package](https://search.nixos.org/packages?channel=23.11&from=0&size=50&sort=relevance&type=packages&query=nodePackages)

However because this project does not have a package-lock.json, this will not work.

Would you like to add a package-lock.json so this can be used by Nix users or addable to the Nix package manager without having any workarounds?

**Comment by rotemdan** [comment]
There is a `package-lock.json` file, but I'm intentionally not including it because users have reported it created problems to them, in another repository, where I did have it in the repository.

I also noticed that the version of `echogarden` itself mentioned in `package-lock.json`, is not up-to-date:

```json
{
	"name": "echogarden",
	"version": "0.11.19",
	"lockfileVersion": 3,
	"requires": true,
	"packages": {
		"": {
			"name": "echogarden",
			"version": "0.11.19",
			"license": "GPL-3.0",
			"os": [
				"win32",
				"darwin",
				"linux"
			],
```

I don't know what is the reasoning for that. I've published two versions after `"0.11.19"` and it hasn't changed.

Based on the answer I got from a chatbot, the lock file, and the version, is only updated when I run `npm install`, not when I run `npm publish`.

It seems wrong for me to post a lock file that includes, an older version of the package. I don't know. If the version number update is not automated when running `npm publish`, then the version number wouldn't stay up-to-date.

`npm` itself doesn't include this file in their registry, but you're saying that somehow it's important to have it on GitHub. I don't know. I'm not an "expert" at this.

**Comment by rotemdan** [comment]
The internal package manager download tar.gz files from Hugging Face.

All the internal packages are versioned based on date.

For each internal package, the exact version of the file is hard-coded into the source code. In that sense it's fully reproducible. That's why I designed it this way:


In `PackageManager.ts`, it resolves the exact version of each internal package (with `20230718` as default).
```ts
const basePackageUrl = "https://huggingface.co/echogarden/echogarden-packages/resolve/main/"

const defaultVersionTag = "20230718"

const packageVersionTagResolutionLookup: { [packageName: string]: string } = {
	"sox-14.4.2-linux-minimal": "20230802",

	"vits-de_DE-thorsten_emotional-medium": "20230808",
	"vits-en_GB-semaine-medium": "20230808",
	"vits-fr_FR-upmc-medium": "20230808",
	"vits-lb_LU-marylux-medium": "20230808",
	"vits-ro_RO-mihai-medium": "20230808",
	"vits-sr_RS-serbski_institut-medium": "20230808",
	"vits-tr_TR-dfki-medium": "20230808",

	"vits-cs_CZ-jirka-medium": "20230824",
	"vits-de_DE-thorsten-high": "20230824",
	"vits-hu_HU-anna-medium": "20230824",
	"vits-pt_PT-tugao-medium": "20230824",
	"vits-sk_SK-lili-medium": "20230824",
	"vits-tr_TR-fahrettin-medium": "20230824",

	"vits-ar_JO-kareem-medium": "20231022",
	"vits-cs_CZ-jirka-low": "20231022",
	"vits-en_US-hfc_male-medium": "20231022",
	"vits-en_US-libritts_r-medium": "20231022",
	"vits-hu_HU-imre-medium": "20231022",
	"vits-pl_PL-mc_speech-medium": "20231022",

	"whisper-tiny": "20231126",
	"whisper-tiny.en": "20231126",
	"whisper-base": "20231126",
	"whisper-base.en": "20231126",
	"whisper-small": "20231126",
	"whisper-small.en": "20231126",
	"whisper-medium": "20231126",
	"whisper-medium.en": "20231126",
	"whisper-large-v3": "20231126",

	"vits-ar_JO-kareem-low": "20231126",
	"vits-en_US-hfc_female-medium": "20231126",
}
```

Whenever I update a package, it is published to a new file, with a different date, so older and newer internal packages coexist on the server.

**Comment by rbozan** [comment]
> There is a `package-lock.json` file, but I'm intentionally not including it because users have reported it created problems to them, in another repository, where I did have it in the repository.
> 
> I also noticed that the version of `echogarden` itself mentioned in `package-lock.json`, is not up-to-date:
> 
> ```json
> {
> 	"name": "echogarden",
> 	"version": "0.11.19",
> 	"lockfileVersion": 3,
> 	"requires": true,
> 	"packages": {
> 		"": {
> 			"name": "echogarden",
> 			"version": "0.11.19",
> 			"license": "GPL-3.0",
> 			"os": [
> 				"win32",
> 				"darwin",
> 				"linux"
> 			],
> ```
> 
> I don't know what is the reasoning for that. I've published 3 versions after `"0.11.19"` and it hasn't changed.
> 
> Based on the answer I got from a chatbot, the lock file, and the version, is only updated when I run `npm install`, not when I run `npm publish`.
> 
> It seems wrong for me to post a lock file that includes, an older version of the package. I don't know. If the version number update is not automated when running `npm publish`, then the version number wouldn't stay up-to-date.
> 
> `npm` itself doesn't include this file in their registry, but you're saying that somehow it's important to have it on GitHub. I don't know. I'm not an "expert" at this.

I think you can consider something like semver-release using GitHub Actions so the bumping of the version in package.json/package-lock.json and doing a new release is completely automated to both GitHub and Npm (https://github.com/semantic-release/semantic-release). This would require you to write commits in conventialcommits style and to make branches instead of directly committing into main.

**Comment by rotemdan** [comment]
In version `0.12.1` I addressed several issues you gave, as well as some additional ones:
* `package-lock.json` is now published on GitHub. I'll try to do my best effort to ensure the version inside it is up to date by making some sort of script for that. I still, though, can't understand how somehow having this file in a GitHub repository, but not on `npm` makes sense. So basically if you install from npm itself, you get no guarantee that the hashes are the same hashes that the developer had when they published it. So what do you do? Only install from GitHub, avoiding using the actual package manager? That's kind of silly.
* Add ffmpeg binaries for many platforms, including pre-signed binaries for macOS (all binaries are now based on binaries from `ffmpeg-static`). Also, in particular there are now also binaries for Linux ARM64. Actually, I believe that now Echogarden should support ARM64 on Linux, but I can't test that.
* Prioritize local `ffmpeg`, and `sox` if on system path - but exclusively use a specific sox package on Windows, since `14.4.2` is broken when reading from stdin. It has to use `14.4.1` instead.
* Add `setGlobalOption` and `getGlobalOption`, with support for overriding the path for `ffmpeg` and `sox` via the API (not in the CLI yet). Next I'll add an option to control the logging verbosity level - that would actually require many different code changes, so it's actually a kind of bigger task than it seems.
* Fixed the annoying `punycode` warning for the CLI by splitting some modules (can't remove it completely since it is caused by a dependency of dependency I have no control of).

I'm not interested in adding extra layers of complexity by using all sorts of release scripts, bundlers, complicated branching and pull request processes, signatures etc. Believe me I tried these things in the past but I have found they mostly my productivity, not increase it.

The package doesn't even have a build script actually, it just compiled via TypeScript, which can be done almost instantly via `tsc` watch - this has made testing and debugging significantly  easier. I can also step into the code in VSCode without running into problems, which has dramatically improved the quality and productivity I'm able to get.

Since I'm developing this by myself, at least for now, it is actually important for me that everything that goes to to GitHub should be working code and already published to npm. So having only a single branch with only published code serves as a kind of quality assurance so that I never publish anything potentially partial or broken. It has worked for me so far.

**Comment by rbozan** [comment]
I have a semi-related question here btw @rotemdan ,

In my production environment I have a container which predownloads `ffmpeg` of the package manager of a Linux distribution so the tool will not download `ffmpeg` at runtime for me.

Is there a way to predownload the other tools beforehand to use them? In my case, I'd like to have `whisper-small-20231126` and `whisper-tiktoken-data-20240408` predownloaded and added to my container, instead of the container downloading these at runtime. 

Cheers.

**Comment by rotemdan** [comment]
Not exactly sure what you mean, so I'll give multiple solutions.

You can run:
```
echogarden install whisper-small-20231126 whisper-tiktoken-data-20240408 ...
```

To fetch and install any number of packages (`uninstall` to remove).

You can also use the storage paths mentioned in [the technical document](https://github.com/echogarden-project/echogarden/blob/main/docs/Technical.md#package-system) to copy the files manually to and from their storage locations (There's nothing else needed other than copying the files with the correct subdirectory name).

Also, I can add a global option to set a custom package storage path. Actually I already did at the last update but then removed it at the last minute since I wasn't sure if it's necessary at this point. I wasn't sure about the use cases. Also, the way the code currently works is based on standard OS directory paths. It will take a slight effort to customize that.

**Comment by rbozan** [comment]
It seems it's exactly what I need:
```
echogarden install whisper-small-20231126 whisper-tiktoken-data-20240408
```

It only doesn't work :( 
```
❯ ./node_modules/.bin/echogarden install whisper-small-20231126                                
Echogarden v1.2.0

Unknown operation: 
```

**Comment by rotemdan** [comment]
It works for me.

Based on your error message, it's likely that other operations shouldn't work as well, like say, `speak`.

Maybe something about the command line / execution is different than normal?

**Comment by rotemdan** [comment]
Maybe try:

```
npx echogarden install whisper-small-20231126 whisper-tiktoken-data-20240408
```

Maybe it's the way it's called? I don't know?

**Comment by rbozan** [comment]
Hm it indeed does not seem to work for any command except just the help ...

```
❯ npx echogarden                    
Echogarden v1.2.0

Supported operations:

echogarden speak text [output files...] [options...]
    Speak the given text

echogarden speak-file inputFile [output files...] [options...]
    Speak the given text file

(... etcetera ...)
```

```
❯ npx echogarden speak
Echogarden v1.2.0

Unknown operation: 
```

Not sure what's up, it works fine using Node.js API though.

**Comment by rotemdan** [comment]
Yes, I now found out that I changed the start up code for the CLI, on the last update (`1.2.0`), in such a way that it only works correctly when a configuration file is given. I'll fix it soon.

Thanks for reporting.

**Comment by rotemdan** [comment]
Okay I quickly published `1.2.1` that should fix it. See if everything's okay.

**Comment by rbozan** [comment]
> Okay I quickly published `1.2.1` that should fix it. See if everything's okay.

Yep all works fine, also the `install` command does what I want. Thanks again.
